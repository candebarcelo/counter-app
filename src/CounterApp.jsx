import { useState } from "react";
import PropTypes from "prop-types";

// each time that there's a change in the state, the whole component code is
// re-run with the new values, and it's re-rendered.
export const CounterApp = ({ value }) => {
  // pass the counter value via setCounter function (decalred here).
  // useState(initialValue).
  const [counter, setCounter] = useState(value);

  const handleAdd = () => {
    // not counter++ bc that needs a 'let', not 'const'.
    setCounter(counter + 1);
  };

  const handleSubstract = () => setCounter(counter - 1);
  const handleReset = () => setCounter(value);

  return (
    // always wrap everything in an empty <> (it's a react Fragment).
    // it doesnt support multiple parent elements.
    <>
      <h1>CounterApp</h1>
      <h2> {counter} </h2>

      <button onClick={handleAdd}> +1 </button>
      <button onClick={handleSubstract}> -1 </button>
      <button aria-label="btn-reset" onClick={handleReset}>
        {" "}
        Reset{" "}
      </button>
    </>
  );
};

// propTypes is to say which are required and of what types.
// there's also defaultTypes to define the default values, it works in the same way.
CounterApp.propTypes = {
  value: PropTypes.number.isRequired,
};
