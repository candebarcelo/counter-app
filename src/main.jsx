import { StrictMode } from "react";
import { createRoot } from "react-dom/client";
import { CounterApp } from "./CounterApp.jsx";
import "./index.css";
import { FirstApp } from "./FirstApp.jsx";

createRoot(document.getElementById("root")).render(
  <StrictMode>
    {/* <FirstApp title="hola" /> */}
    <CounterApp value={60}></CounterApp>
  </StrictMode>
);
