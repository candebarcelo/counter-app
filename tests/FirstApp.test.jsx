import { render, screen } from "@testing-library/react";
import { FirstApp } from "../src/FirstApp";

describe("FirstApp", () => {
  const title = "Hellooo";
  const subTitle = "I'm a subtitle";

  test("should match with snapshot", () => {
    const { container } = render(<FirstApp title={title} />);

    // to make sure html structure doesnt change. u can
    // update the snapshot if u need to change it. it's useful for
    // components that should almost never change.
    expect(container).toMatchSnapshot();
  });

  test("should show correct text", () => {
    render(<FirstApp title={title} />);
    // screen.debug() // <-- u can use this to see the current rendered html in the console.
    expect(screen.getByText(title)).toBeTruthy();
  });

  test("should show title as h1", () => {
    render(<FirstApp title={title} />);
    expect(screen.getByRole("heading", { level: 1 }).innerHTML).toContain(
      title
    );

    //another way to do it:
    // const { getByText, getByTestId } = render(<FirstApp title={title} />);

    // expect(getByText(title)).toBeTruthy();

    // // const h1 = container.querySelector("h1");
    // // expect(h1.innerHTML).toContain(title);

    // expect(getByTestId("test-title").innerHTML).toContain(title);
  });

  test("should show subtitle sent via props", () => {
    render(<FirstApp title={title} subTitle={subTitle} />);

    expect(screen.getAllByText(subTitle).length).toBe(2);

    // another way of doing it:
    // const { getAllByText } = render(
    //   <FirstApp title={title} subTitle={subTitle} />
    // );

    // expect(getAllByText(subTitle).length).toBe(2);
  });
});
