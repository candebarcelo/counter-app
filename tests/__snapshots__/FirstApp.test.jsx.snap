// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FirstApp should match with snapshot 1`] = `
<div>
  <h1
    data-testid="test-title"
  >
     
    Hellooo
     
  </h1>
  <p>
    No hay subtítulo
  </p>
  <p>
    No hay subtítulo
  </p>
  <p>
    Fernando Herrera
  </p>
</div>
`;
